### PR TITLE
Connect to Minio with temporary credentials from STS obtained with OpenID Connect token from Keycloak

### LINKED ISSUE #10917: Expired temporary credentials (STS) should be refreshed automatically
> **884512a** created the issue

Not requiring program restart
___
Attachments

* [`Cyberduck-creds-expired.png`](https://github-cyberduck-assets.s3.amazonaws.com/10917/Cyberduck-creds-expired.png) (23.6 KiB)

### LINKED ISSUE #13377: aws sso based bookmarks are too difficult to use
I stopped using Cyberduck a couple years ago as multi-account became increasingly common, and I moved to use of aws-vault to store credentials in the MacOS or Windows keychain or use AWS SSO (pre-control tower), neither of which worked with Cyberduck, making it increasingly difficult to use without keeping a duplicate set of credentials.

I decided to try it again today, and noticed the new credentials option to obtain credentials from ~/.aws/config, where I have a bunch of SSO based credentials setup for many accounts organized via AWS Control Tower. For the first account, I added the bookmark, specified the profile name (based on SSO), and then double-clicked to open - nothing.

I figured, just as with the command line, I needed to first run "aws sso login --profile $profile", which I did, then also confirmed on the same CLI I could see data using a simple command like "aws ec2 describe-regions --profile $profile", so when I went back into Cyberduck and then clicked on the bookmark, it worked. I was glad to see this appeared to make Cyberduck usable again for large-scale / enterprise-level AWS multi-account scenarios!

So, I then went in and setup a bunch of other bookmarks based on other SSO profiles. But when I next tried to double-click these profiles, they would all hang. I went into the CLI, confirmed the profile worked, then once at least one CLI command was run against the profile (note I was never prompted for another aws sso login command on the CLI once the first one was done) on the CLI, Cyberduck would then work against that profile.  OK, this is not good, I may have dozens of accounts, so this is going to make this so inconvenient again, just like before, to where the cons outweigh the pros...

I Googled your docs and found a reference https://docs.cyberduck.io/protocols/s3/ where it indicates both the aws sso login and at least one command like aws sts get-caller-identity needs to be run to get the STS tokens you need into cache for use. So, at least with the current design, I see what's happening.

The problem is, this effectively makes Cyberduck difficult to use, so much so I'm unlikely to want to use it much still. So, I thought I'd suggest a couple of improvements which you could make to reduce developer friction on use.

1. When I setup the bookmark, using the "S3 (Credentials from AWS Command Line Interface)" type, you have the name of the profile in the ~/.aws/config file, which should be readable by your program, so you can look into that profile and detect that it's based on SSO. Why do I have to run the "aws sts get-caller-identity --profile $profile"? Why can't you run this for me, some exec method to run a shell command as the user?
2. Same as #1, if you detect that this is an SSO profile, if you try to run #1, and that fails with an error code indicating the user has not logged in yet, why can't you run the "aws sso login --profile $profile" command for me? This does interact with the browser, not sure if that complicates things, and if so this does only need to be run once, so it's less onerous, but if you can't run it, at least popping up a dialog box saying "Open a terminal and run 'aws sso login --profile $profile', then retry" or similar.

The problem is #1 when you have a large number of accounts. It makes this program less usable than it should/could be. running a couple of shell commands when you have every parameter of data needed to run them from the configuration you collect would seem to be trivial, but would result in significant ease-of-use improvements. If you can't do this, at least explain why, and use dialog boxes to indicate to the user what they must do, so we don't have to guess or search your documentation for why this doesn't "just work".

### LINKED ISSUE #13804: Suport to login using temporary credentials from Security Token Service (STS API) using OpenID Connect web identity
I considered writing this on #13377 or #13381, however these are mostly related to AWS specifically which isn't quite what I am looking for. I am currently using Wasabi for S3 and looking for a way to authenticate users from an IdP (SSO). 

Wasabi supports _AssumeRole_, however to use that directly I would need to create users on Wasabi and manage their privileges there; and I would rather manage user accounts and their privileges from our IdP.

My goals are:

1. Avoid users having to use anything other than Cyberduck. Users should be able to install a profile and be prompted to login.
1. Configuring privileges (roles) is handled outside of Cyberduck.
1. Should be compatible with anything that returns credentials in a valid format, e.g. `AssumeRoleWithWebIdentityResponse` or `AssumeRoleResponse`

I would like to do something like the following:

1. Cyberduck opens a browser window to a configurable URL (as it does for Dropbox/GDrive).
1. User authenticates in the browser window to an external service.
1. External service uses Wasabi's STS to create appropriately privileged temporary credentials.
1. Callback to Cyberduck with ??? (see below)

I am not entirely sure how the callback to Wasabi could be done or what is feasible. The Cyberduck would obviously need the credentials, and some way to refresh the credentials when they are about to expire, like a session token or cookie.

While AWS' S3 SDK is very general and allows access to any compliant service, their authentication, particularly related to SSO seems to be very locked to their IdP. Additionally, using an external configuration file (e.g. `~/.aws/credentials`) is not user friendly.

I've read through the documentation and not found anything helpful, and neither are other issues on the topic such as this one: #11133

### PR DESCRIPTION
**Pull request by SheromS**
This pull request contains our current state with:

- Authentication to Keycloak
- Fetch STS temporary credentials
- Testcontainers with the included docker-compose configuration
- First tests including the testcontainers (are running)
- A first version of the OIDC Cyberduck Profile in the ressource-folder
- Suggested changes from @chenkins implemented (about STS) https://github.com/SheromS/cyberduck/pull/6
- rebase on [#14882](https://github.com/iterate-ch/cyberduck/pull/14882). This allows to properly sign requests after refreshing tokens.

- [x] Resolves #13804
- [ ] Resolves #10917

- [x] Make sure we have test coverage for expired STS tokens.
- [x] Resolves TODOs
- [x] Resolve comments

### CODE REVIEW

**Comment by dkocher** [comment]
Somehow this is based on 10045f45 from bugfix/GH-10917-authentication-interceptor but contains newer commits from `master`. Please rebase again.

File: s3/src/test/resources/S3-OIDC-Testing.cyberduckprofile [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        <key>STS Endpoint</key>
+        <string>http://localhost:9000</string>
+        <key>Scopes</key>
+        <array>
+            <string>openid</string>
+            <string>profile</string>
+            <string>minio-authorization</string>
+        </array>
+        <key>Password Configurable</key>
+        <false/>
```
Comment (dkocher) [inline s3/src/test/resources/S3-OIDC-Testing.cyberduckprofile:57]: Because `PasswordGrant` is used, this should be `true`.

File: s3/src/test/resources/S3-OIDC-Testing.cyberduckprofile [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        <key>Username Configurable</key>
+        <false/>
```
Comment (dkocher) [inline s3/src/test/resources/S3-OIDC-Testing.cyberduckprofile:59]: Because `PasswordGrant` is used, this should be `true`.

File: s3/src/main/java/ch/cyberduck/core/s3/S3WebIdentityTokenExpiredResponseInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            if(log.isWarnEnabled()) {
+                log.warn(String.format("Skip retry for response %s after %d executions", response, executionCount));
+            }
+        }
+        return false;
+    }
+
+    private void refreshOAuthAndSTS() {
+        try {
+            if(host.getCredentials().getOauth().isExpired()) {
```
[the remark this answers is not in the document]
Comment (dkocher) [reply s3/src/main/java/ch/cyberduck/core/s3/S3WebIdentityTokenExpiredResponseInterceptor.java:102]: Either always refresh the OAuth tokens or make it dependent on the error response code if possible.

File: s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                        this.save(this.refresh(tokens));
+                    }
+                    catch(InteroperabilityException | LoginFailureException e3) {
+                        log.warn(String.format("Failure %s refreshing OAuth tokens", e3));
+                        try {
+                            this.save(this.authorize(host, prompt, new DisabledCancelCallback()));
+                        }
+                        catch(BackgroundException e) {
+                            log.warn(String.format("Failure %s OAuth authentication", e));
+                            throw new BackgroundException(e);
```
Comment (dkocher) [inline s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java:89]: Why is `e` re-wrapped?

File: s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        this.key = key;
+        this.prompt = prompt;
+        this.session = session;
+    }
+
+    @Override
+    public void process(final org.apache.http.HttpRequest request, final HttpContext context) throws HttpException, IOException {
+        if(System.currentTimeMillis() >= stsExpiryInMilliseconds) {
+            try {
+                if(tokens.isExpired()) {
```
Comment (dkocher) [inline s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java:78]: It should be possible to just call `super.process` here to refresh the OAuth tokens.

File: s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        return credentials;
+    }
+
+    private AWSSecurityTokenService getTokenService(final Host host) {
+        final ClientConfiguration configuration = new CustomClientConfiguration(host,
+                new ThreadLocalHostnameDelegatingTrustManager(trust, host.getProtocol().getSTSEndpoint()), key);
+        return AWSSecurityTokenServiceClientBuilder
+                .standard()
+                .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(host.getProtocol().getSTSEndpoint(), null))
+                .withCredentials(new AWSCredentialsProvider() {
```
Comment (dkocher) [inline s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java:164]: Replace with `new StaticCredentialsProvider(new AnonymousAWSCredentials())`

**Review by dkocher — CHANGES_REQUESTED** [review]
- [ ] Review comments.
- [x] Please run code formatting for all new classes to match existing code style.
- [x] Change package name of test classes to `sts`?

File: oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java [hunk 7]
```diff
@@ -135,25 +136,27 @@ public OAuthTokens authorize(final Host bookmark, final LoginCallback prompt, fi
         if(log.isDebugEnabled()) {
             log.debug(String.format("Start new OAuth flow for %s with missing access token", bookmark));
         }
-        final TokenResponse response;
+        // Save access key and refresh key
         switch(flowType) {
             case AuthorizationCode:
-                response = this.authorizeWithCode(bookmark, prompt);
-                break;
+                final IdTokenResponse authorizationCodeResponse = this.authorizeWithCode(bookmark, prompt);
```
[the remark this answers is not in the document]
Comment (SheromS) [reply oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java]: good point. I added the id token on every occurence and used it in tests `getIdToken()`. I also added it in the class `DefaultHostPasswordStore`. I hope this is good.

File: s3/src/main/java/ch/cyberduck/core/s3/S3WebIdentityTokenExpiredResponseInterceptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (SheromS) [reply s3/src/main/java/ch/cyberduck/core/s3/S3WebIdentityTokenExpiredResponseInterceptor.java]: Now it always refreshes the token in case of these specific s3 error codes which makes sense.

File: s3/src/main/java/ch/cyberduck/core/s3/S3WebIdentityTokenExpiredResponseInterceptor.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                try {
+                    final S3ServiceException failure;
+                    if(null != response.getEntity()) {
+                        EntityUtils.updateEntity(response, new BufferedHttpEntity(response.getEntity()));
+                        failure = new S3ServiceException(response.getStatusLine().getReasonPhrase(),
+                                EntityUtils.toString(response.getEntity()));
+                    }
+                    // minio sometimes packs the error code and description in the http header
+                    else{
+                        failure = new S3ServiceException(response.getStatusLine().getReasonPhrase());
```
[the remark this answers is not in the document]
Comment (SheromS) [reply s3/src/main/java/ch/cyberduck/core/s3/S3WebIdentityTokenExpiredResponseInterceptor.java:71]: The InvalidAccessKeyId error code has no response body.
I made a commented [test](https://github.com/SheromS/cyberduck/blob/9eb90698c2c46557fb48139fc4e17266bfbd749b/s3/src/test/java/ch/cyberduck/core/oidc_sts/OidcAuthenticationTest.java#L151C11-L151C11) which shows an edge case scenario. The request leaves the app right before the expiry time is reached. Because of the network latency the expiry time is reach at the minio service. The AccessKey got just outdated and the Failure gets handled in this class `S3WebIdentityTokenExpiredResponseInterceptor`.
What are your thoughts about that?

File: s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java (hunk 5 above)
Comment (SheromS) [reply s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java:77]: To me this is not possible because in the super refresh there is no catch for the specific exceptions to do an oAuth authorization afterwards. Additionally, it sets a wrong AUTHORIZATION Header which is not a problem because this Header is recalculated in an interceptor later in the chain but to me it is not nice.

**Comment by SheromS** [comment]
- Code formatting for all new classes is done.
- Currently the name of the testing package is oidc_sts. If you wish I change it to sts but in `OidcAuthenticationTest` there are some tests only considering the oidc part.

**Review by dkocher — CHANGES_REQUESTED** [review]
- [x] Change package name of new STS related production and test classes to `ch.cyberduck.core.sts`. While OIDC is part of the specific flow, this could be replaced by other `assumeRole` implementations using LDAP or other identity providers.

File: oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java [hunk 9]
```diff
@@ -136,21 +135,24 @@ public OAuthTokens authorize(final Host bookmark, final LoginCallback prompt, fi
...
+        // Save access token, refresh token and id token
         switch(flowType) {
             case AuthorizationCode:
-                final IdTokenResponse authorizationCodeResponse = this.authorizeWithCode(bookmark, prompt);
-                return credentials.withToken(authorizationCodeResponse.getIdToken()).withOauth(new OAuthTokens(
-                        authorizationCodeResponse.getAccessToken(), authorizationCodeResponse.getRefreshToken(),
-                        null == authorizationCodeResponse.getExpiresInSeconds() ? System.currentTimeMillis() :
-                                System.currentTimeMillis() + authorizationCodeResponse.getExpiresInSeconds() * 1000, authorizationCodeResponse.getIdToken()))
+                response = this.authorizeWithCode(bookmark, prompt);
+                return credentials.withToken(response.getIdToken()).withOauth(new OAuthTokens(
```
[the remark this answers is not in the document]
Comment (SheromS) [reply oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java]: You are right. This is not required anymore.

File: s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java (hunk 5 above)
Comment (SheromS) [reply s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java:77]: I definitely see the point of avoiding code duplication. I don't see how to solve that with the `OAuth2ErrorResponseInterceptor`. This is an `ServiceUnavailableRetryStrategy` and reacts on 401 SC Response but we decided to refresh the tokens preemptively in an interceptor.
I thought about extracting the code to a separate method but there are always tiny differences that lead to incompatibility.

File: s3/src/main/java/ch/cyberduck/core/s3/S3WebIdentityTokenExpiredResponseInterceptor.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (SheromS) [reply s3/src/main/java/ch/cyberduck/core/s3/S3WebIdentityTokenExpiredResponseInterceptor.java]: We took your [changes](https://github.com/iterate-ch/cyberduck/pull/14893/files/3528d70556808a5c1c61ad2bd37ec3b5fd0b10dd#r1264642467) with the minio headers. Is that fine for now?

File: oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java [hunk 10]
```diff
@@ -345,13 +354,14 @@ else if(value instanceof Number) {
...
-            return new TokenResponse()
+        public IdTokenResponse toTokenResponse() {
+            return new IdTokenResponse()
                     .setTokenType(tokenType)
                     .setScope(scope)
                     .setExpiresInSeconds(expiresInSeconds)
                     .setAccessToken(accessToken)
-                    .setRefreshToken(refreshToken);
+                    .setRefreshToken(refreshToken)
+                    .setIdToken(idToken);
```
Comment (dkocher) [inline oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java:364]: This causes a `NullPointerException` when no id token is returned.

Comment (SheromS) [reply oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java]: Good point. Is resolved by the latest commit.

File: core/src/main/java/ch/cyberduck/core/DefaultHostPasswordStore.java [hunk 11]
```diff
@@ -272,6 +279,10 @@ public void delete(final Host bookmark) throws LocalAccessDeniedException {
             if(credentials.getOauth().getExpiryInMilliseconds() != null) {
                 this.deletePassword(this.getOAuthHostname(bookmark), String.format("%s OAuth2 Token Expiry", prefix));
             }
+            if(StringUtils.isNotBlank(credentials.getOauth().getRefreshToken())) {
```
Comment (dkocher) [inline core/src/main/java/ch/cyberduck/core/DefaultHostPasswordStore.java:282]: Check should be on `getIdToken`

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -31,3 +31,12 @@ jobs:
       - name: Publish Test Report
         if: ${{ always() }}
         uses: ScaCap/action-surefire-report@v1.7.0
+
```
Comment (dkocher) [inline .github/workflows/ci.yml:34]: Indent seems to be wrong making the file syntax invalid.

File: core/src/main/java/ch/cyberduck/core/DefaultHostPasswordStore.java [hunk 13]
```diff
@@ -232,6 +234,11 @@ public void save(final Host bookmark) throws LocalAccessDeniedException {
                 this.addPassword(this.getOAuthHostname(bookmark), String.format("%s OAuth2 Token Expiry", prefix),
                         String.valueOf(credentials.getOauth().getExpiryInMilliseconds()));
             }
+            if(StringUtils.isNotBlank(credentials.getOauth().getIdToken())) {
+                this.addPassword(bookmark.getProtocol().getScheme(),
+                        bookmark.getPort(), this.getOAuthHostname(bookmark),
+                        String.format("%s OIDC Id Token", prefix), credentials.getOauth().getRefreshToken());
```
Comment (dkocher) [inline core/src/main/java/ch/cyberduck/core/DefaultHostPasswordStore.java:240]: Saving refresh token instead of id token.

File: oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java [hunk 14]
```diff
@@ -135,25 +135,30 @@ public OAuthTokens authorize(final Host bookmark, final LoginCallback prompt, fi
...
         }
-        final TokenResponse response;
+
+        final IdTokenResponse response;
+        // Save access token, refresh token and id token
         switch(flowType) {
             case AuthorizationCode:
                 response = this.authorizeWithCode(bookmark, prompt);
-                break;
+                return credentials.withOauth(new OAuthTokens(
```
Comment (dkocher) [inline oauth/src/main/java/ch/cyberduck/core/oauth/OAuth2AuthorizationService.java:144]: It has been discussed before that these two return statement are identical and can be moved below the switch.

File: s3/src/test/java/ch/cyberduck/core/sts/OidcAuthenticationTest.java [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+        // assert that STS credentials expires with the oAuth token even though the duration seconds is valid for longer
+        STSCredentialsRequestInterceptor authorizationService = session.getAuthorizationService();
+        assertTrue(40 > ((authorizationService.getStsExpiryInMilliseconds() - System.currentTimeMillis()) / 1000));
+        assertEquals(Optional.of(authorizationService.getStsExpiryInMilliseconds()).get() / 1000, authorizationService.getTokens().getExpiryInMilliseconds() / 1000);
+
+        session.close();
+    }
+
+    /** only use with the below specified changes in the keycloak config json file and run as separate test
```
Comment (dkocher) [inline s3/src/test/java/ch/cyberduck/core/sts/OidcAuthenticationTest.java:120]: Ideally we find a way to run these tests with different configurations in `keycloak-realm.json`.

Comment (IsmCadi) [reply s3/src/test/java/ch/cyberduck/core/sts/OidcAuthenticationTest.java:120]: My first thought was to use a second realm with these configurations, but that would lead to a second IDP config in MinIO which is not possible with the same address. As of now I do not know how to do this without maybe using additional containers, which would complicate the Testcontainers and the Docker config.

Comment (chenkins) [reply s3/src/test/java/ch/cyberduck/core/sts/OidcAuthenticationTest.java:120]: @dkocher I have drafted a way to run test with different configurations in `keycloak-realm.json`.
`testSTSCredentialsExpiredValidOAuthToken` will take 15 minutes as the min STS duration is 900s see https://min.io/docs/minio/linux/developers/security-token-service/AssumeRoleWithWebIdentity.html. So unsure what it makes sense to finish it.

@IsmCadi I struggle to see the  working of `testBucketRequestBeforeTokenExpiryFailsBecauseOfLatency` - can you re-phrase what you're testing and how to adjust the 28 secs? The test is red when I run locally.

Comment (chenkins) [reply s3/src/test/java/ch/cyberduck/core/sts/OidcAuthenticationTest.java:120]: Let's continue discussion here: https://github.com/iterate-ch/cyberduck/pull/14972/files#r1291199162

**Comment by chenkins** [comment]
Closing this pr after discussion with @dkocher -  superseded by #14972 .
